sklearn/tests/test_pipeline.py

```diff
@@ -928,3 +948,210 @@ def test_make_pipeline_memory():
     assert_true(pipeline.memory is None)
 
     shutil.rmtree(cachedir)
+
+
+def test_column_transformer():
+    # dictionary
+    X_dict = {'first': np.array([0, 1, 2]),
+              'second': np.array([2, 4, 6])}
+    # recarray
+    X_recarray = np.recarray((3,),
+                             dtype=[('first', np.int), ('second', np.int)])
+    X_recarray['first'] = X_dict['first']
+    X_recarray['second'] = X_dict['second']
+
+    # array
+    X_array = np.array([[0, 1, 2], [2, 4, 6]]).T
+
+    Xs_name = [X_dict, X_recarray]
+    Xs_positional = [X_array]
+
+    # dataframe
+    try:
+        import pandas as pd
+        X_df = pd.DataFrame(X_dict)
+        Xs_name.append(X_df)
+        Xs_positional.append(X_df)
+    except:
+        print("Pandas not found, not testing ColumnTransformer with"
+              " DataFrame.")
+
+    X_res_first1D = np.array(X_dict['first'])
+    X_res_first2D = X_res_first1D.reshape(-1, 1)
+    X_res_both = np.vstack([X_dict['first'], X_dict['second']]).T
+
+    for X in Xs_name:
+        col_trans = ColumnTransformer([('trans', Trans(), 'first')])
+        assert_array_equal(col_trans.fit_transform(X), X_res_first1D)
+        assert_array_equal(col_trans.fit(X).transform(X), X_res_first1D)
+
+        col_trans = ColumnTransformer([('trans', Trans(), ['first'])])
+        assert_array_equal(col_trans.fit_transform(X), X_res_first2D)
+        assert_array_equal(col_trans.fit(X).transform(X), X_res_first2D)
+
+        col_trans = ColumnTransformer([('trans', Trans(),
+                                        ['first', 'second'])])
+        assert_array_equal(col_trans.fit_transform(X), X_res_both)
+        assert_array_equal(col_trans.fit(X).transform(X), X_res_both)
+
+        col_trans = ColumnTransformer([('trans1', Trans(), ['first']),
+                                       ('trans2', Trans(), ['second'])])
+        assert_array_equal(col_trans.fit_transform(X), X_res_both)
+        assert_array_equal(col_trans.fit(X).transform(X), X_res_both)
+
+        if hasattr(X, 'iloc'):
+            # for dataframes also test slice
+            col_trans = ColumnTransformer(
+                [('trans', Trans(), slice('first', 'second'))])
+            assert_array_equal(col_trans.fit_transform(X), X_res_both)
+            assert_array_equal(col_trans.fit(X).transform(X), X_res_both)
```

Comment:
Discussed in person. Summary: this follows pandas semantics (since it slices the pandas object under the hood) to include the end label of the slice (in case of using labels, when slicing positional with integers, it follows typical python/numpy semantics to not include end)